---
title: "RNAseq"
output:
  html_document:
    self_contained: true
    toc: true
    toc_float: true
    toc-location: left
editor: visual
---

## Setup

Load packages

```{r warning=FALSE, message=FALSE, results = FALSE}
library(data.table)
library(stringr)
library(ggplot2)
theme_py <- theme_light() + theme(
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA),
  text = element_text(size=20),
  strip.placement = "outside", 
  strip.text = element_text(size=20, color="black"),
  strip.background = element_rect(fill="white")
)
theme_set(theme_py)
library(patchwork)
library(ggrepel)
library(ComplexHeatmap)
library(Rsubread)
library(DESeq2)
library(caret)
library(pROC)
```

Directories and inputs

```{r}
dat_dir <- "gene_counts"
res_dir <- file.path(dat_dir, "results")
fig_dir <- file.path(dat_dir, "plots")
for (newdir in c(res_dir, fig_dir))
  dir.create(newdir, showWarnings = FALSE)
```

## Mapping QCs

```{r}
parse_flagstat <- function(flagstat_file, pattern = "properly paired") {
  lns <- readLines(flagstat_file)
  ln <- grep(pattern, lns, value = TRUE)[1]
  as.integer(stringr::str_extract(ln, "\\d+"))
}
samples <- fread(file.path(dat_dir, "design.tsv"))$sample
flag_fns <- sapply(samples, function(smp) {file.path(
    smp,
    "alignments",
    sprintf("%s.Aligned.sortedByCoord.out.bam.flagstat", smp)
)}, USE.NAMES = TRUE)
flag_dt <- data.table(
  sample = names(flag_fns)
)
flag_dt[, total_reads := sapply(
    flag_dt$filename, parse_flagstat, "total"
)]
flag_dt[, mapped_reads := sapply(
    flag_dt$filename, parse_flagstat, "mapped"
)]
flag_dt[, dup_reads := sapply(
    flag_dt$filename, parse_flagstat, "secondary"
)]
flag_dt[, percent_mapped := mapped_reads / total_reads]
flag_dt[, percent_dups := dup_reads / total_reads]
flag_dt
fwrite(flag_dt, file.path(res_dir, "mapping_qc.tsv"), sep = "\t")
```

## Gene counts

Load design table

```{r}
col_fn <- file.path(dat_dir, "design.tsv")
col_dt <- fread(col_fn)
col_dt[, molecule := factor(molecule, levels = c("control", "cApUp", "cGAMP"))]
col_dt[, biol_rep := factor(biol_rep, levels = 1:3)]
col_dt[, tech_rep := factor(tech_rep, levels = c("A", "B", "C"))]

mol_cols <- c("cApUp" = "#ff7f00", "cGAMP" = "#984ea3", "control" = "#4daf4a")

col_df <- copy(col_dt)
class(col_df) <- "data.frame"
rownames(col_df) <- col_df$sample
```

Load gene counts from all samples into expression matrix

```{r eval=FALSE, include=FALSE}
samples <- col_dt$sample
con_lst <- sapply(samples, function(smp) {
    fn <- list.files(
        file.path(smp, "alignments"),
        "*ReadsPerGene.out.tab$",
        full.names = TRUE
    )
    dt <- fread(fn, skip = 4L)[, 1:2]
    setnames(dt, c("gene", smp))
    dt
}, USE.NAMES = TRUE, simplify = FALSE)
con_dt <- Reduce(function(...) merge(..., by = "gene", all = TRUE), con_lst)
fwrite(con_dt, file.path(dat_dir, "raw_counts_rnaseq.tsv"), sep = "\t")
```

```{r}
samples <- col_dt$sample
con_lst <- sapply(samples, function(smp) {
    fn <- list.files(
        file.path(smp, "alignments"),
        "*.featureCounts.paired.tab$",
        full.names = TRUE
    )
    dt <- fread(fn, select = c(1, 7), header = TRUE)
    setnames(dt, c("gene", smp))
    dt
}, USE.NAMES = TRUE, simplify = FALSE)
con_dt <- Reduce(function(...) merge(..., by = "gene", all = TRUE), con_lst)
fwrite(con_dt, file.path(dat_dir, "raw_counts_rnaseq.tsv"), sep = "\t")
```

Load count matrix

```{r}
con_fn <- file.path(dat_dir, "raw_counts_rnaseq.tsv")
con_dt <- fread(con_fn, header = TRUE)
con_mt <- data.matrix(con_dt[, -1])
con_mt <- con_mt[, rownames(col_df)]
rownames(con_mt) <- con_dt[[1]]
```

Inspect TPMs distribution

```{r fig.width=8, fig.height=6, warning=FALSE, message=FALSE}
# load gene lengths
gtf <- as.data.table(
    rtracklayer::import("genome/Spis_long.annot.gtf")
)
gen_lens <- structure(gtf$width, names = gtf$gene_id)
# calculate TPMs
tpm <- function(counts, len) {
  x <- counts / len
  return(t(t(x) * 1e6 / colSums(x)))
}
tpm_mt <- tpm(con_mt, gen_lens[rownames(con_mt)])
tpm_dt <- as.data.table(tpm_mt, keep.rownames = "gene")
tpm_dt <- melt.data.table(
  tpm_dt,
  id.vars = "gene",
  variable.name = "sample",
  value.name = "TPM"
)
tpm_dt <- merge.data.table(tpm_dt, col_dt, by = "sample")
tpm_dt[, sample := factor(sample, levels = unique(tpm_dt$sample))]

gp_exp <- ggplot(tpm_dt, aes(sample, log10(TPM), fill = molecule)) +
  geom_violin(scale = "width", alpha = 0.8, color = "black") +
  geom_boxplot(width = 0.5, outlier.shape = NA, alpha = 0.8, color = "black") +
  scale_fill_manual(values = mol_cols, limits = force) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.01))) +
  labs(x = "samples", y = "log10(TPMs)") +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
    legend.title = element_blank(), legend.position = "none"
  )

var_dt <- tpm_dt[, .(var = var(TPM)), .(gene)]
gp_var <- ggplot(var_dt, aes(log10(var))) + 
  geom_density() +
  scale_x_continuous(limits = c(-4,NA)) +
  scale_y_continuous(expand = expansion(mult = c(0,0.01))) +
  labs(x = "log10(expression variance)")

gp_pch <- gp_exp / gp_var + plot_layout(heights = c(2,1))
gp_pch
```

```{r include=FALSE, eval=TRUE}
pdf(file.path(fig_dir, "tpms_var.pdf"), width=8, height=6)
gp_pch
dev.off()
```

Data for differential expression analysis

```{r}
# DESeq2 object
require(DESeq2)
col_df$sample <- paste(col_df$molecule, col_df$biol_rep, sep = "_")
dds <- DESeqDataSetFromMatrix(
  countData = con_mt,
  colData = col_df,
  design = ~ molecule
)
saveRDS(dds, file.path(res_dir, "dds.rds"))

# collapse technical replicates
ddscol <- collapseReplicates(dds, dds$sample)
saveRDS(ddscol, file.path(res_dir, "ddscol.rds"))
```

Use normalized log-transformed expression data

```{r}
# normalize samples
dds <- readRDS(file.path(res_dir, "ddscol.rds"))
dds <- estimateSizeFactors(dds)
norm_mt <- counts(dds, normalized = TRUE)

# row normalize to bring genes to same range
norm_mt <- (norm_mt + 10) / apply(norm_mt + 10, 1, median) 
norm_mt <- log2(norm_mt)

# save
write.table(
  norm_mt,
  file.path(res_dir, "norm_expression.tsv"),
  sep = "\t",
  row.names = TRUE,
  quote = FALSE
)
```

## PCA

PCA on collapsed samples

```{r fig.width=8, fig.height=10, warning=FALSE}
set.seed(1950)
pca_res <- prcomp(t(norm_mt), center = TRUE)

# variance explained
pca_var <- data.table(
    pct_var = round(((pca_res$sdev) ^ 2 /
    sum((pca_res$sdev) ^ 2)* 100), 2)
)
pca_var[, pct_cum := cumsum(pct_var)]
pca_var[, PC := factor(seq_len(.N))]
gp_var <- ggplot(pca_var, aes(PC, pct_var)) + 
  geom_bar(stat = "identity") +
  geom_line(aes(y = pct_cum, group = 1)) + 
  geom_point(aes(y = pct_cum)) +
  scale_y_continuous(expand = expansion(0.01,0)) +
  labs(y = "% of variance\nexplained", x = "PC") +
  theme(panel.grid.major.y = element_line(size = 0.5))

pca_dt <- as.data.table(pca_res$x, keep.rownames = "sample")
pca_dt <- merge.data.table(col_df, pca_dt, by = "sample", sort = FALSE)

gp_bip <- ggplot(pca_dt, aes(PC1, PC2, fill = molecule, shape = biol_rep)) +
  geom_point(size = 5) +
  scale_fill_manual(values = mol_cols) +
  scale_shape_manual(values = c("1" = 21, "2" = 24)) +
  guides(fill = guide_legend(override.aes = list(shape = 21)))

gp_var / gp_bip
```


```{r include=FALSE, eval=TRUE}
pdf(file.path(fig_dir, "pca_collapsed.pdf"), width = 6, height = 8)
gp_var / gp_bip
dev.off()
```

## Marker genes

Use normalized gene counts

```{r}
# load counts
norm_mt <- read.table(
  file.path(res_dir, "norm_expression.tsv"),
  header = TRUE
)
# load column data
col_dt <- fread(col_fn)
col_dt[, sample := paste(molecule, biol_rep, sep = "_")]
col_dt[, tech_rep := NULL]
col_dt <- unique(col_dt)
```

Identify marker genes

```{r include=TRUE, eval=TRUE}
# gene markers by normalized expression
genes_fc <- names(which(apply(norm_mt, 1, function(x) {
  sort(x, decreasing = TRUE)[1] >= 2
})))
genes_vari <- names(which(apply(norm_mt, 1, function(x) {
  var(x) > 1.5
})))
```

```{r include=TRUE, eval=TRUE}
# gene markers by high FC + significant DEseq2 LTR test
dds <- readRDS(file.path(res_dir, "ddscol.rds"))
dds <- DESeq(dds, test = "LRT", reduced = ~ 1)
dds_res <- results(dds)
dds_qval <- dds_res$padj
names(dds_qval) <- rownames(dds_res)
genes_deseq <- names(which(dds_qval < 1e-2))

genes_high <- names(which(apply(norm_mt, 1, function(x) {
  sort(x, decreasing = TRUE)[2] > 1.5
})))

gene_marks <- intersect(genes_high, genes_deseq)
```

Load gene sets of interest

```{r}
cglr_genes <- readLines(file.path("annotation", "cGLR.txt"))
cglr_genes <- paste0("Spis_", cglr_genes)
cglr_genes <- str_replace(cglr_genes, "\\.", "_")
sting_genes <- readLines(file.path("annotation", "STING.txt"))
sting_genes <- paste0("Spis_", sting_genes)
sting_genes <- str_replace(sting_genes, "\\.", "_")

gene_marks <- unique(c(
  gene_marks,
  cglr_genes[cglr_genes %in% rownames(norm_mt)],
  sting_genes[sting_genes %in% rownames(norm_mt)]
))
```

Select number of clusters for genes

```{r fig.width=10, fig.height=6}
set.seed(1950)

# determine k for kmeans
ks <- 1:30
tot_withinss <- sapply(ks,  function(k) {
  cl <- kmeans(norm_mt[gene_marks, ], k)
  cl$tot.withinss
})
elbow_dt <- data.table(k = ks, tot_withinss = tot_withinss)
elbow_gp <- ggplot(elbow_dt, aes(x = k, y = tot_withinss)) +
  geom_line() + geom_point() +
  scale_x_continuous(breaks = ks) + 
  theme(panel.grid.major = element_line(size = 0.5))
elbow_gp
```

```{r include=FALSE, eval=TRUE}
pdf(file.path(fig_dir, "kmeans_elbow.pdf"), width = 8, height = 6)
elbow_gp
dev.off()
```

Cluster genes

```{r}
set.seed(1950)

# kmeans
k <- 10
cl <- kmeans(norm_mt[gene_marks, ], k)
gene_order_list <- tapply(names(cl$cluster), cl$cluster, function(gs) {
  cor_genes <- cor(t(norm_mt[gs,]))
  hclust_genes <- hclust(as.dist(1 - cor(cor_genes)), method = "ward.D2")
  rownames(cor_genes)[hclust_genes$order]
})
names(gene_order_list) <- unique(cl$cluster)
gene_order_list <- gene_order_list[as.character(seq_along(gene_order_list))]

# cluster clusters
cluster_order <- hclust(
  dist(cor(t(cl$centers)),
  method = "euclidean"),
  method = "ward.D2"
)$order
gene_order_list <- gene_order_list[as.character(cluster_order)]
gene_order <- unname(unlist(gene_order_list[cluster_order]))
clusters_dt <- data.table(
  gene = unlist(gene_order_list),
  clusters = as.character(
    rep(names(gene_order_list), sapply(gene_order_list, length))
  )
)

# group clusters (manually)
clusters_dt[, group := as.integer(clusters)]
clusters_dt[, group := factor(
  group,
  levels = c(1, 4, 5, 8, 6, 7, 3, 2, 10, 9)
)]
clusters_dt[, group := as.integer(group)]
setorder(clusters_dt, group)
gene_order <- clusters_dt$gene

# save
fwrite(clusters_dt, file.path(res_dir, "genes_clusters.tsv"), sep = "\t")
```

Heatmap of markers

```{r fig.height=10, fig.width=8, warning=FALSE, message=FALSE}
# order rows and columns
samples_order <- col_dt$sample
plot_mt <- norm_mt[gene_order, samples_order]

# center at 0
plot_min <- quantile(abs(range(plot_mt)), 0.75)
plot_min <- 4
plot_mt[plot_mt < (-1 * plot_min)] <- -1 * plot_min
plot_mt[plot_mt > plot_min] <- plot_min

# heatmap colors
col_vec <- colorRampPalette(RColorBrewer::brewer.pal(11, 'BrBG'))(1000)
col_fun <- circlize::colorRamp2(
  seq(-plot_min, plot_min, length.out = length(col_vec)),
  col_vec
)

# color annotations
col_ann <- HeatmapAnnotation(
    which = "column",
    border = TRUE,
    "molecule" = as.character(
      col_dt[match(colnames(plot_mt), sample)]$molecule
    ),
    col = list("molecule" = mol_cols)
)

# gene module annotations
grann <- clusters_dt[match(rownames(plot_mt), gene)]$group
grann_lab <- unique(clusters_dt[match(rownames(plot_mt), gene)]$group)
grann <- factor(grann, levels = grann_lab)
module_ann <- HeatmapAnnotation(
    which = "row", border = TRUE,
    "group" = anno_block(
      labels = grann_lab,
      gp = gpar(col = NA)
    )
)
row_split <- grann

# genes annotations
gene_ann <- fread(
    file.path("annotation", "Spis_annotation_v7_2020-10-26_clean"),
    select = 1:3
)
tf_ann <- fread(
  file.path("annotation", "curated_TFh_Spis_EDITED_NAMES.csv"),
  header = FALSE
)[, c(1,3,2)]
setnames(gene_ann, c("gene", "pfam", "og"))
setnames(tf_ann, c("gene", "pfam", "og"))
gene_ann <- rbindlist(list(
  tf_ann, gene_ann[!gene %in% tf_ann$gene]
))
gene_ann[, name := ""]
gene_ann[og != "", name := paste(gene, og, sep = " | ")]
gene_ann[
  nchar(name) > 70,
  name := paste0(substr(name, 1, 67), "...")
]

marks <- gene_ann[gene %in% rownames(plot_mt)][name != ""]
row_labels_marks_ids <- match(marks$gene, rownames(plot_mt))
row_labels_marks <- marks[
  match(rownames(plot_mt)[row_labels_marks_ids], gene)
]$name

gset_marks <- sapply(rownames(plot_mt), function(x) {
  if (x %in% cglr_genes) {
    "cGLR"
  } else if (x %in% sting_genes) {
    "STING"
  } else {
    " "
  }
})
pval_marks <- sapply(rownames(plot_mt), function(x) {
  if (x %in% genes_deseq) {
    "*"
  } else {
    ""
  }
})

mark_ann <- HeatmapAnnotation(
    which = "row", border = TRUE,
    "gene set" = anno_simple(
      gset_marks,
      col = c("cGLR" = "#e6ab02", "STING" = "#3690c0", " " = "#d9d9d9"),
      pch = pval_marks
    ),
    "marker" = anno_mark(
      at = row_labels_marks_ids,
      labels = row_labels_marks, labels_gp = gpar(fontsize = 8)
    )
)
lgd_set <- Legend(
  title = "gene\nset",
  at = c("cGLR", "STING", " "),
  legend_gp = gpar(fill = c("#e6ab02", "#3690c0", "#d9d9d9")),
)
lgd_sig <- Legend(
  title = "pval<0.01",
  pch = "*",
  type = "points",
  labels = "< 0.01"
)

# heatmap
hm <- Heatmap(
  plot_mt, name = "normalized\nexpression",
  col = col_fun,
  cluster_rows = FALSE, cluster_columns = FALSE,
  show_row_names = FALSE, show_column_names = TRUE,
  row_title = "genes",
  row_split = row_split,
  cluster_row_slices = FALSE,
  top_annotation = col_ann,
  right_annotation = mark_ann,
  left_annotation = module_ann,
  border = TRUE
)
draw(hm, annotation_legend_list = list(lgd_set, lgd_sig))
```

```{r include=FALSE, eval=TRUE}
pdf(
    file.path(fig_dir, "clustering_heatmap.pdf"),
    width = 8, height = 32
)
draw(hm, annotation_legend_list = list(lgd_set, lgd_sig))
dev.off()
```