---
title: "Comparison"
output:
  html_document:
    self_contained: true
    toc: true
    toc_float: true
    toc-location: left
editor: visual
---

## Setup

```{r warning=FALSE, message=FALSE, results = FALSE}
library(data.table)
library(stringr)
library(ggplot2)
theme_py <- theme_light() + theme(
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA),
  text = element_text(size = 20),
  strip.placement = "outside",
  strip.text = element_text(size = 20, color = "black"),
  strip.background = element_rect(fill = "white")
)
theme_set(theme_py)
library(patchwork)
library(eulerr)
source("functions.R")
```

Directories and inputs

```{r}
spis_dir <- file.path(
    "Stylophora_pistillata", "differential_expression", "results"
)
cvig_dir <- file.path(
    "Crassostrea_virginica", "differential_expression", "results"
)
data_dir <- file.path("comparison")
figs_dir <- file.path(data_dir, "plots")
for (newdir in c(figs_dir))
  dir.create(newdir, showWarnings = FALSE)
```

Load ortholog pairs

```{r}
og <- fread(file.path(
    data_dir, "orthologous_pairs.txt.gz"
), header = FALSE)
og <- unique(rbindlist(list(
    og[grepl("Cvig", V1) & grepl("Spis", V2)][, 2:1],
    og[grepl("Spis", V1) & grepl("Cvig", V2)]
), use.names = FALSE))
setnames(og, c("gene_spis", "gene_cvig"))
```

Load Stylophora annotations

```{r}
gene_ann <- fread(file.path(
    "Stylophora_pistillata", "annotation", "Spis_annotation_v7_2020-10-26_clean"
), select = 1:3)
tf_ann <- fread(file.path(
    "Stylophora_pistillata", "annotation", "curated_TFh_Spis_EDITED_NAMES.csv"
), header = FALSE
)[, c(1, 3, 2)]
setnames(gene_ann, c("gene", "pfam", "og"))
setnames(tf_ann, c("gene", "pfam", "og"))
gene_ann <- rbindlist(list(
  tf_ann, gene_ann[!gene %in% tf_ann$gene]
))
gene_ann[, name := ""]
gene_ann[og != "", name := paste(gene, og, sep = " | ")]
gene_ann[
  nchar(name) > 70,
  name := paste0(substr(name, 1, 67), "...")
]
setnames(gene_ann, "gene", "gene_spis")
```

Overlap DE gene lists from both species

```{r}
spis_res <- fread(file.path(spis_dir, "dss_res.tsv"))
cvig_res <- fread(file.path(cvig_dir, "dss_res.tsv"))

# map cvig gene IDs to protein IDs
list_int_fixd <- dictionary_t2g(
  gtf_fn = file.path(
    "Crassostrea_virginica", "genome",
    "GCF_002022765.2_C_virginica-3.0_genomic.gtf"
  ),
  vector_to_fix = str_remove(cvig_res$gene, "Cvig_"),
  t2g = FALSE,
  transcript_field = "CDS",
  transcript_id = "protein_id",
  gene_id = "gene_id"
)
list_int_prot <- list_int_fixd[!is.na(names(list_int_fixd))]
list_int_prot <- structure(
  paste0("Cvig_", list_int_prot),
  names = paste0("Cvig_", names(list_int_prot))
)
cvig_res[, gene_id := gene][, gene := list_int_prot[gene_id]]
cvig_res <- cvig_res[!is.na(gene)]
cvig_res[, gene_id := NULL]

# significant genes
padj_thr <- 0.05
lfc_thr <- 1
cvig_sgn <- cvig_res[log2FoldChange > lfc_thr & padj < padj_thr]
spis_sgn <- spis_res[log2FoldChange > lfc_thr & padj < padj_thr]
spis_sgn <- spis_sgn[gene %in% spis_sgn[, .N, gene][N == 2]$gene]
genes <- list(
  "spis" = spis_sgn$gene,
  "cvig" = cvig_sgn$gene
)

# select spis genes upregulated in intersect
spis_sgn <- spis_res[gene %in% genes$spis]
setnames(spis_sgn, colnames(spis_sgn), paste0(colnames(spis_sgn), "_spis"))
setnames(spis_sgn, "treatment_spis", "treatment")

# select cvig genes upregulated in intersect
cvig_sgn <- cvig_res[gene %in% genes$cvig]
setnames(cvig_sgn, colnames(cvig_sgn), paste0(colnames(cvig_sgn), "_cvig"))
setnames(cvig_sgn, "treatment_cvig", "treatment")

# combine
og_sgn <- merge.data.table(
  merge.data.table(og, spis_sgn, by = "gene_spis"),
  merge.data.table(og, cvig_sgn, by = "gene_cvig"),
  by = c("gene_spis", "gene_cvig", "treatment")
)
setorder(og_sgn, treatment, log2FoldChange_spis)
unique(og_sgn[, .(gene_spis, gene_cvig)])

# add spis gene annotations
og_sgn_ann <- merge.data.table(
  og_sgn, gene_ann, by = "gene_spis", all.x = TRUE, sort = FALSE
)
setcolorder(og_sgn_ann, c("gene_cvig", "gene_spis", "pfam", "og"))
og_sgn_ann[, c("name") := NULL]

# save
fwrite(
  og_sgn_ann,
  file.path(data_dir, sprintf("dds_padj%s_ogs.tsv", padj_thr)),
  sep = "\t"
)
```

Venn diagram

```{r}
og_sgn_all <- merge.data.table(
  merge.data.table(
    og, unique(spis_sgn[, .(gene_spis)]), by = "gene_spis", all.y = TRUE
  ),
  merge.data.table(
    og, unique(cvig_sgn[, .(gene_cvig)]), by = "gene_cvig", all.y = TRUE
  ),
  by = c("gene_spis", "gene_cvig"), all = TRUE
)


class(og_sgn_all) <- "data.frame"
og_sgn_all[!is.na(og_sgn_all)] <- 1
og_sgn_all[is.na(og_sgn_all)] <- 0
og_sgn_all[, 1] <- as.integer(og_sgn_all[, 1])
og_sgn_all[, 2] <- as.integer(og_sgn_all[, 2])
fit_eu <- euler(og_sgn_all)

eul <- plot(
  fit_eu,
  quantities = TRUE,
  main = "significantly upregulated genes"
)
eul
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(figs_dir, "venn_ogs.pdf"),
  width = 5, height = 4
)
eul
dev.off()
```

```{r}
marks <- fread(
  file.path(data_dir, "Cvig_Spis_OG_labels_shortlist.txt"),
  fill = TRUE
)
setnames(marks, c("gene_spis", "pfam", "name"))
marks_all <- merge.data.table(marks, og, all.x = TRUE)
marks_all[, label := NULL]

marks_spis <- merge.data.table(marks_all, spis_sgn, by = "gene_spis")
marks_spis[, label := NULL]
setnames(
  marks_spis,
  colnames(marks_spis)[-1],
  str_remove(colnames(marks_spis)[-1], "_spis")
)

marks_cvig <- merge.data.table(marks_all, cvig_sgn, by = "gene_cvig")
setnames(
  marks_cvig,
  colnames(marks_cvig)[-1],
  str_remove(colnames(marks_cvig)[-1], "_cvig")
)

marks_fcs <- rbindlist(list(
  spis = marks_spis,
  cvig = marks_cvig
), idcol = "species", use.names = TRUE)
marks_fcs[, minuslog10padj := -1 * log10(padj)]
marks_fcs[, species_treatment := paste(species, treatment, sep = "_")]
marks_fcs[, sample := switch(species_treatment,
  spis_cApUp = "S. pistillata cApUp",
  spis_cGAMP = "S. pistillata cGAMP",
  cvig_cApUp = "C. virginica cApUp",
  cvig_cGAMP = "C. virginica cGAMP"
), by = seq_along(marks_fcs$treatment)]
marks_fcs[, sample := factor(sample, levels = c(
  "S. pistillata cApUp",
  "S. pistillata cGAMP",
  "C. virginica cApUp",
  "C. virginica cGAMP"
))]
marks_fcs[, id := sprintf("%s | %s", gene_spis, gene_cvig)]
marks_fcs[, id := sprintf("%s | %s", name, pfam)]
marks_top <- marks_fcs[order(-log2FoldChange)][, .SD[1], .(sample, gene_spis)]

marks_dc <- dcast.data.table(
  marks_top[, .(id, sample, log2FoldChange)],
  id ~ sample,
  value.var = "log2FoldChange"
)
marks_mt <- marks_dc[, -1]
rownames(marks_mt) <- marks_dc[[1]]
marks_hc <- hclust(dist(marks_mt), method = "ward.D2")

marks_top[, id := factor(id, levels = rownames(marks_mt)[rev(marks_hc$order)])]
setorder(marks_top, id)

gp_og <- ggplot(
  marks_top,
  aes(sample, id, size = minuslog10padj, fill = log2FoldChange)
) +
  geom_point(shape = 21) +
  labs(
    x = "gene", y = "",
    size = "log2\nfold change",
    fill = "-log10\nadjusted p value"
  ) +
  scale_y_discrete(
    breaks = marks_top$id,
    labels = marks_top$name
  ) +
  scale_size_continuous(
    range = c(1, 10),
    breaks = c(10, 30, 50)
  ) +
  scale_fill_distiller(
    direction = 1,
    breaks = c(0, 2, 4),
    limits = c(0, 5), oob = scales::squish
  ) +
  theme(
    panel.grid.major = element_line(linewidth = 0.2),
    axis.text.x = element_text(angle = 60, vjust = 1, hjust = 1),
    axis.ticks = element_blank()
  )

gp_nm <- ggplot(unique(marks_top[, .(id, pfam)]), aes("x", id)) +
  scale_y_discrete(
    position = "left",
    breaks = marks_top$id,
    labels = marks_top$pfam
  ) +
  scale_x_discrete(
    expand = c(0, 0)
  ) +
  theme(
    axis.line.x = element_blank(),
    axis.text.x = element_blank(),
    axis.text.y = element_text(hjust = 0),
    axis.ticks = element_blank(),
    axis.title = element_blank(),
    plot.margin = unit(c(0, 0, 0, 0), "null"),
    panel.spacing = unit(c(0, 0, 0, 0), "null"),
    panel.border = element_blank()
  )

gp_h <- (gp_og  + gp_nm) +
  plot_layout(widths = c(3, 1), guides = "collect")
gp_h
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(figs_dir, "dot_ogs.pdf"),
  width = 18, height = 5
)
gp_h
dev.off()
```