---
title: "Crassostrea virginica"
output:
  html_document:
    self_contained: true
    toc: true
    toc_float: true
    toc-location: left
editor: visual
---

## Setup

Load packages

```{r warning=FALSE, message=FALSE, results = FALSE}
library(data.table)
library(stringr)
library(ggplot2)
theme_py <- theme_light() + theme(
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA),
  text = element_text(size = 20),
  strip.placement = "outside",
  strip.text = element_text(size = 20, color = "black"),
  strip.background = element_rect(fill = "white")
)
theme_set(theme_py)
library(patchwork)
library(ggrepel)
library(RColorBrewer)
library(ComplexHeatmap)
library(Rsubread)
library(DESeq2)
library(eulerr)
library(pathview)
library(KEGGREST)
library(gage)
library(openxlsx)
source("functions.R")
```

Directories and inputs

```{r}
dat_dir <- file.path("Crassostrea_virginica", "differential_expression")
ann_dir <- file.path("Crassostrea_virginica", "annotation")
gen_dir <- file.path("Crassostrea_virginica", "genome")
res_dir <- file.path(dat_dir, "results")
fig_dir <- file.path(dat_dir, "plots")
for (newdir in c(res_dir, fig_dir))
  dir.create(newdir, showWarnings = FALSE)

mol_cols <- c("cApUp" = "#ff7f00", "cGAMP" = "#984ea3", "control" = "#4daf4a")
```

Gene annotations

```{r}
gene_ann <- fread(file.path(ann_dir, "gene_info.tsv"))
gene_ann[, gene := paste0("Cvig_", gene_name)]
gene_ann[gene_description == "- && - && -", gene_description := ""]
gene_ann[tf_family == "-0", tf_family := ""]
gene_ann <- gene_ann[, .(gene, gene_id, gene_description, tf_family)]

# proteins
list_int_fixd <- dictionary_t2g(
  gtf_fn = file.path(
    "Crassostrea_virginica", "genome",
    "GCF_002022765.2_C_virginica-3.0_genomic.gtf"
  ),
  vector_to_fix = str_remove(gene_ann$gene, "Cvig_"),
  t2g = FALSE,
  transcript_field = "CDS",
  transcript_id = "protein_id",
  gene_id = "gene_id"
)
list_int_prot <- list_int_fixd[!is.na(names(list_int_fixd))]
list_int_prot <- structure(
  paste0("Cvig_", list_int_prot),
  names = paste0("Cvig_", names(list_int_prot))
)
gene_ann[, gene_id := gene][, gene := list_int_prot[gene_id]]
gene_ann <- gene_ann[is.na(gene), gene := gene_id]
gene_ann <- gene_ann[, .(gene, gene_id, gene_description, tf_family)]

# ogs
og <- fread(file.path(
    "comparison", "orthologous_pairs.txt.gz"
), header = FALSE, skip = 1L)
og <- unique(rbindlist(list(
    og[grepl("Cvig", V1) & grepl("Spis", V2)][, 2:1],
    og[grepl("Spis", V1) & grepl("Cvig", V2)]
), use.names = FALSE))
setnames(og, c("gene_spis", "gene"))
gene_ann_og <- merge.data.table(
  gene_ann, og, by = "gene", all.x = TRUE
)

# spis ann
spis_ann <- fread(file.path(
    "Stylophora_pistillata", "annotation", "Spis_annotation_v7_2020-10-26_clean"
), select = 1:3)
setnames(spis_ann, c("gene_spis", "pfam", "og"))
gene_ann <- merge.data.table(
  gene_ann_og, spis_ann, by = "gene_spis", all.x = TRUE
)
fwrite(gene_ann, file.path(ann_dir, "gene_ann.tsv"), sep = "\t")

# gene sets of interest
cglr_genes <- readLines(file.path(ann_dir, "cGLR.txt"))
cglr_genes <- paste0("Cvig_", cglr_genes)
cglr_genes <- str_replace(cglr_genes, "\\.", "_")
sting_genes <- readLines(file.path(ann_dir, "STING.txt"))
sting_genes <- paste0("Cvig_", sting_genes)
sting_genes <- str_replace(sting_genes, "\\.", "_")
```
```{r eval=FALSE, include=FALSE}
# orthologs
og <- fread("comparison/orthologous_pairs.txt", header = FALSE)
og <- unique(rbindlist(list(
    og[grepl("Cvig", V1) & grepl("Spis", V2)][, 2:1],
    og[grepl("Spis", V1) & grepl("Cvig", V2)]
), use.names = FALSE))

# spis genes
cglr_genes <- readLines(file.path(
    "Stylophora_pistillata", "annotation", "cGLR.txt"
))
cglr_genes <- paste0("Spis_", cglr_genes)
cglr_genes <- str_replace(cglr_genes, "\\.", "_")

sting_genes <- readLines(file.path(
    "Stylophora_pistillata", "annotation", "STING.txt"
))
sting_genes <- paste0("Spis_", sting_genes)
sting_genes <- str_replace(sting_genes, "\\.", "_")

cglr_genes %in% og$V1
sting_genes %in% og$V1
```

## Gene counts

Load design table

```{r}
col_fn <- file.path(dat_dir, "design.tsv")
col_dt <- fread(col_fn)
col_dt[, treatment := factor(
  treatment, levels = c("control", "cApUp", "cGAMP")
)]
col_dt[, biol_rep := factor(biol_rep, levels = c(6, 12))]
col_dt[, tech_rep := factor(tech_rep, levels = 1:4)]

col_df <- copy(col_dt)
class(col_df) <- "data.frame"
rownames(col_df) <- col_df$sample
```

Load count matrix

```{r}
con_fn <- file.path(dat_dir, "gene_count.tsv")
con_dt <- fread(con_fn, header = TRUE)
con_mt <- data.matrix(con_dt[, -1])
con_mt <- con_mt[, rownames(col_df)]
rownames(con_mt) <- con_dt[[1]]
```

Inspect TPMs distribution

```{r fig.width=8, fig.height=6, warning=FALSE, message=FALSE}
# load gene lengths
ginfo <- fread(file.path(ann_dir, "gene_info.tsv"))
ginfo[, gene_name := paste0("Cvig_", gene_name)]
ginfo[, width := abs(gene_end - gene_start)]
gen_lens <- structure(ginfo$width, names = ginfo$gene_name)

# calculate TPMs
tpm <- function(counts, len) {
  x <- counts / len
  return(t(t(x) * 1e6 / colSums(x)))
}
tpm_mt <- tpm(con_mt, gen_lens[rownames(con_mt)])
tpm_dt <- as.data.table(tpm_mt, keep.rownames = "gene")
tpm_dt <- melt.data.table(
  tpm_dt,
  id.vars = "gene",
  variable.name = "sample",
  value.name = "TPM"
)
tpm_dt <- merge.data.table(tpm_dt, col_dt, by = "sample")
tpm_dt[, sample := factor(sample, levels = unique(tpm_dt$sample))]

gp_exp <- ggplot(tpm_dt, aes(sample, log10(TPM), fill = treatment)) +
  geom_violin(scale = "width", alpha = 0.8, color = "black") +
  geom_boxplot(width = 0.5, outlier.shape = NA, alpha = 0.8, color = "black") +
  scale_fill_manual(values = mol_cols, limits = force) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.01))) +
  labs(x = "samples", y = "log10(TPMs)") +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
    legend.title = element_blank(), legend.position = "none"
  )

var_dt <- tpm_dt[, .(var = var(TPM)), .(gene)]
gp_var <- ggplot(var_dt, aes(log10(var))) + 
  geom_density() +
  scale_x_continuous(limits = c(-4,NA)) +
  scale_y_continuous(expand = expansion(mult = c(0,0.01))) +
  labs(x = "log10(expression variance)")

gp_pch <- gp_exp / gp_var + plot_layout(heights = c(2,1))
gp_pch
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "tpms_var.pdf"),
  width = 10, height = 6
)
gp_pch
dev.off()
```

Data for differential expression analysis

```{r}
# DESeq2 object
require(DESeq2)
dds <- DESeqDataSetFromMatrix(
  countData = con_mt,
  colData = col_df,
  design = ~ treatment
)
saveRDS(dds, file.path(res_dir, "dds.rds"))

# collapse technical replicates
# col_df$sample <- paste(col_df$treatment, col_df$biol_rep, sep = "_")
# ddscol <- collapseReplicates(dds, dds$sample)
# saveRDS(ddscol, file.path(res_dir, "ddscol.rds"))
```

Use normalized log-transformed expression data

```{r}
# normalize samples
dds <- readRDS(file.path(res_dir, "dds.rds"))
dds <- estimateSizeFactors(dds)
norm_mt <- counts(dds, normalized = TRUE)

# row normalize to bring genes to same range
norm_mt <- (norm_mt + 10) / apply(norm_mt + 10, 1, median)
norm_mt <- log2(norm_mt)

# save
write.table(
  norm_mt,
  file.path(res_dir, "norm_expression.tsv"),
  sep = "\t",
  row.names = TRUE,
  quote = FALSE
)
```

## Samples correlation

```{r fig.width=7, fig.height=6}
dds <- readRDS(file.path(res_dir, "dds.rds"))
vsd <- vst(dds, blind = FALSE)
mark_genes <- names(which(apply(norm_mt, 1, function(x) max(abs(x)) > 2)))
smp_dist <- dist(t(assay(vsd)[mark_genes, ]))
smp_dist_mt <- as.matrix(smp_dist)
smp_cor_mt <- cor(smp_dist_mt)

#ord <- rownames(smp_cor_mt)[hclust(dist(smp_cor_mt), method = "single")$order]
#ord <- rownames(smp_cor_mt)[c(1:12, 15, 13:14)]
#smp_cor_mt <- smp_cor_mt[ord, ord]

col_vec <- rev(colorRampPalette(rev(brewer.pal(9, "RdBu")))(255))
col_fun <- circlize::colorRamp2(
  seq(-1, 1, length.out = length(col_vec)),
  col_vec
)

col_ann <- HeatmapAnnotation(
    which = "column",
    border = TRUE,
    "treatment" = as.character(
      col_dt[match(colnames(smp_cor_mt), sample)]$treatment
    ),
    col = list("treatment" = mol_cols)
)

row_ann <- HeatmapAnnotation(
    which = "row",
    border = TRUE,
    show_legend = FALSE,
    "treatment" = as.character(
      col_dt[match(colnames(smp_cor_mt), sample)]$treatment
    ),
    col = list("treatment" = mol_cols)
)

smp_dist_hm <- Heatmap(
  smp_cor_mt,
  name = "sample\ncorrelation", col = col_fun,
  cluster_rows = FALSE, cluster_columns = FALSE,
  top_annotation = col_ann, left_annotation = row_ann,
  border = TRUE
)
smp_dist_hm
```

```{r include=FALSE, eval=TRUE}
pdf(
    file.path(fig_dir, "samples_dist_all_samples.pdf"),
    width = 7, height = 5.6
)
smp_dist_hm
dev.off()
```

## PCA

PCA on (collapsed) samples

```{r fig.width=8, fig.height=10, warning=FALSE}
set.seed(1950)
pca_res <- prcomp(t(norm_mt), center = TRUE)

# variance explained
pca_var <- data.table(
    pct_var = round(((pca_res$sdev) ^ 2 /
    sum((pca_res$sdev) ^ 2) * 100), 2)
)
pca_var[, pct_cum := cumsum(pct_var)]
pca_var[, PC := factor(seq_len(.N))]
gp_var <- ggplot(pca_var, aes(PC, pct_var)) +
  geom_bar(stat = "identity") +
  geom_line(aes(y = pct_cum, group = 1)) +
  geom_point(aes(y = pct_cum)) +
  scale_y_continuous(expand = expansion(0.01, 0)) +
  labs(y = "% of variance\nexplained", x = "PC") +
  theme(panel.grid.major.y = element_line(size = 0.5))

pca_dt <- as.data.table(pca_res$x, keep.rownames = "sample")
pca_dt <- merge.data.table(col_dt, pca_dt, by = "sample", sort = FALSE)

gp_bip <- ggplot(pca_dt, aes(PC1, PC2, fill = treatment, shape = biol_rep)) +
  geom_point(size = 5) +
  scale_fill_manual(values = mol_cols) +
  scale_color_manual(values = mol_cols) +
  scale_shape_manual(values = c("6" = 21, "12" = 24)) +
  geom_text_repel(aes(color = treatment, label = sample)) +
  guides(fill = guide_legend(override.aes = list(shape = 21)))

gp_var / gp_bip
```


```{r include=FALSE, eval=TRUE}
pdf(file.path(fig_dir, "pca.pdf"), width = 8, height = 10)
gp_var / gp_bip
dev.off()
```

## Marker genes

Use normalized gene counts

```{r}
# load counts
norm_mt <- read.table(
  file.path(res_dir, "norm_expression.tsv"),
  header = TRUE
)
colnames(norm_mt) <- str_replace_all(colnames(norm_mt), "\\.", "-")
# load column data
col_dt <- fread(col_fn)
```

Identify marker genes

```{r include=TRUE, eval=TRUE}
# gene markers by normalized expression
genes_fc <- names(which(apply(norm_mt, 1, function(x) {
  sort(x, decreasing = TRUE)[1] >= 2
})))
genes_vari <- names(which(apply(norm_mt, 1, function(x) {
  var(x) > 1.5
})))
```

```{r include=TRUE, eval=TRUE}
# gene markers by high FC + significant DEseq2 LTR test
dds <- readRDS(file.path(res_dir, "dds.rds"))
dds <- DESeq(dds, test = "LRT", reduced = ~ 1)
dds_res <- results(dds)
dds_qval <- dds_res$padj
names(dds_qval) <- rownames(dds_res)
genes_deseq <- names(which(dds_qval < 1e-2))

genes_high <- names(which(apply(norm_mt, 1, function(x) {
  sort(x, decreasing = TRUE)[2] > 1.5
})))

gene_marks <- intersect(genes_high, genes_deseq)
```

Select number of clusters for genes

```{r fig.width=10, fig.height=6}
set.seed(1950)

# determine k for kmeans
ks <- 1:30
tot_withinss <- sapply(ks,  function(k) {
  cl <- kmeans(norm_mt[gene_marks, ], k)
  cl$tot.withinss
})
elbow_dt <- data.table(k = ks, tot_withinss = tot_withinss)
elbow_gp <- ggplot(elbow_dt, aes(x = k, y = tot_withinss)) +
  geom_line() + geom_point() +
  scale_x_continuous(breaks = ks) + 
  theme(panel.grid.major = element_line(size = 0.5))
elbow_gp
```

```{r include=FALSE, eval=TRUE}
pdf(
    file.path(fig_dir, "kmeans_elbow.pdf"),
    width = 10, height = 6)
elbow_gp
dev.off()
```

Cluster genes

```{r}
set.seed(1950)

# kmeans
k <- 8
cl <- kmeans(norm_mt[gene_marks, ], k)
gene_order_list <- tapply(names(cl$cluster), cl$cluster, function(gs) {
  cor_genes <- cor(t(norm_mt[gs,]))
  hclust_genes <- hclust(as.dist(1 - cor(cor_genes)), method = "ward.D2")
  rownames(cor_genes)[hclust_genes$order]
})
names(gene_order_list) <- unique(cl$cluster)
gene_order_list <- gene_order_list[as.character(seq_along(gene_order_list))]

# cluster clusters
cluster_order <- hclust(
  dist(cor(t(cl$centers)),
  method = "euclidean"),
  method = "ward.D2"
)$order
gene_order_list <- gene_order_list[as.character(cluster_order)]
gene_order <- unname(unlist(gene_order_list[cluster_order]))
clusters_dt <- data.table(
  gene = unlist(gene_order_list),
  clusters = as.character(
    rep(names(gene_order_list), sapply(gene_order_list, length))
  )
)

# group clusters (manually)
clusters_dt[, group := as.integer(clusters)]
clusters_dt[, group := factor(
  group
  #levels = c(1, 4, 5, 8, 6, 7, 3, 2, 10, 9)
)]
clusters_dt[, group := as.integer(group)]
setorder(clusters_dt, group)
gene_order <- clusters_dt$gene

# gene annotaiton
clusters_dt[, gene := factor(gene, levels = gene_order)]
setorder(clusters_dt, gene)
clusters_dt <- merge.data.table(
  clusters_dt, gene_ann,
  by = "gene", all.x = TRUE, sort = FALSE
)

# save
clusters_dt[, clusters := NULL]
clusters_dt[, name := NULL]
fwrite(clusters_dt, file.path(res_dir, "genes_clusters.tsv"), sep = "\t")
```

Heatmap of markers

```{r fig.height=10, fig.width=8, warning=FALSE, message=FALSE}
# order rows and columns
samples_order <- col_dt[sample %in% colnames(norm_mt)]$sample
plot_mt <- norm_mt[gene_order, samples_order]

# center at 0
plot_min <- quantile(abs(range(plot_mt)), 0.75)
plot_min <- 5
plot_mt[plot_mt < (-1 * plot_min)] <- -1 * plot_min
plot_mt[plot_mt > plot_min] <- plot_min

# heatmap colors
col_vec <- colorRampPalette(RColorBrewer::brewer.pal(11, "BrBG"))(1000)
col_fun <- circlize::colorRamp2(
  seq(-plot_min, plot_min, length.out = length(col_vec)),
  col_vec
)

# color annotations
col_ann <- HeatmapAnnotation(
    which = "column",
    border = TRUE,
    "treatment" = as.character(
      col_dt[match(colnames(plot_mt), sample)]$treatment
    ),
    col = list("treatment" = mol_cols)
)

# gene module annotations
grann <- clusters_dt[match(rownames(plot_mt), gene)]$group
grann_lab <- unique(clusters_dt[match(rownames(plot_mt), gene)]$group)
grann <- factor(grann, levels = grann_lab)
module_ann <- HeatmapAnnotation(
    which = "row", border = TRUE,
    "group" = anno_block(
      labels = grann_lab,
      gp = gpar(col = NA)
    )
)
row_split <- grann

# genes annotations
marks <- gene_ann[gene %in% rownames(plot_mt)][gene_description != ""]
marks[, name := strtrim(strsplit(
    gene_description, " && "
)[[1]][1], 50), by = seq_len(nrow(marks))]
marks <- marks[!grepl("ncharacterized LOC", name) & name != "-"]
row_labels_marks_ids <- match(marks$gene, rownames(plot_mt))
row_labels_marks <- marks[
  match(rownames(plot_mt)[row_labels_marks_ids], gene)
]$name

mark_ann <- HeatmapAnnotation(
    which = "row", border = TRUE,
    "marker" = anno_mark(
      at = row_labels_marks_ids,
      labels = row_labels_marks, labels_gp = gpar(fontsize = 5)
    )
)

# heatmap
hm <- Heatmap(
  plot_mt, name = "normalized\nexpression",
  col = col_fun,
  cluster_rows = FALSE, cluster_columns = FALSE,
  show_row_names = FALSE, show_column_names = TRUE, 
  column_names_gp = gpar(fontsize = 8),
  row_title = "genes",
  row_split = row_split,
  cluster_row_slices = FALSE,
  top_annotation = col_ann,
  right_annotation = mark_ann,
  left_annotation = module_ann,
  border = TRUE
)
draw(hm)
```

```{r include=FALSE, eval=TRUE}
pdf(
    file.path(fig_dir, "clustering_heatmap.pdf"),
    width = 8, height = 96
)
draw(hm)
dev.off()
```

## Differential gene analysis

Load design table

```{r}
col_fn <- file.path(dat_dir, "design.tsv")
col_dt <- fread(col_fn)
col_dt[, treatment := factor(
  treatment, levels = c("control", "cApUp", "cGAMP")
)]
col_dt[, biol_rep := factor(biol_rep, levels = c(6, 12))]
col_dt[, tech_rep := factor(tech_rep, levels = 1:4)]

mol_cols <- c("cApUp" = "#ff7f00", "cGAMP" = "#984ea3", "control" = "#4daf4a")

col_df <- copy(col_dt)
class(col_df) <- "data.frame"
rownames(col_df) <- col_df$name
```

Load count matrix

```{r}
con_fn <- file.path(dat_dir, "gene_count.tsv")
con_dt <- fread(con_fn, header = TRUE)
con_mt <- data.matrix(con_dt[, -1])
con_mt <- con_mt[, rownames(col_df)]
rownames(con_mt) <- con_dt[[1]]
```

Data for differential expression analysis

```{r}
# DESeq2 object
require(DESeq2)
dds <- DESeqDataSetFromMatrix(
  countData = con_mt,
  colData = col_df,
  design = ~ treatment
)
dds <- DESeq(dds)
saveRDS(dds, file.path(res_dir, "dds.rds"))
```

Contrasts to control

```{r}
dds <- readRDS(file.path(res_dir, "dds.rds"))
padj_thr <- 0.05
lfc_thr <- 1

# label immune genes
label_genes <- rbindlist(list(
  gene_ann[grep("NFkB|NACHT|IRF|STAT\\d+", pfam, ignore.case = TRUE)][
    , name := str_replace_all(pfam, c(
      "IRF3/IRF4/IRF5/IRF6/IRF7/IRF8/IRF9" = "IRF3/4/5/6/7/8/9",
      "NFKB1/NFKB2/REL/RELA/RELB" = "NFKB1/2",
      "STAT5A/STAT5B/STAT6" = "STAT5/6",
      "STAT_int/STAT_alpha/STAT_bind/SH2/" = "STA5A",
      "RHD_DNA_bind/RHD_dimer/Ank_2/Ank_4/Death/" = "NFKB1"
    ))
  ],
  gene_ann[grep(
    "GN=NFKB|GN=Irf|STAT\\d+",
    gene_description,
    ignore.case = TRUE
  )][, name := str_extract(gene_description, "(?<=GN=)[A-z0-9]+")]
))

label_genes[
  grepl("HEPN_DZIP3/NACHT/NOD2_WH/LRR_6", pfam),
  name := "NLR HEPN_DZIP3/NACHT/NOD2_WH/LRR_6/"
]
label_genes[
  grepl("HEPN_DZIP3/NACHT/LRR_6/", pfam),
  name := "NLR HEPN_DZIP3/NACHT/LRR_6/"
]
label_genes[
  grepl("DUF4062/NACHT/WD40/", pfam),
  name := "NLR DUF4062/NACHT/WD40/"
]
label_genes[
  grepl("DUF4062/NACHT/", pfam),
  name := "NLR DUF4062/NACHT/"
]
label_genes[, gene := str_replace(gene, "\\.", "_")]
label_genes_immuno_1 <- structure(label_genes$name, names = label_genes$gene)

# label immune genes from stylophora
immune_dt <- fread(
  file.path("Stylophora_pistillata", "annotation", "immune.txt"),
  header = FALSE
)
setnames(immune_dt, "V1", c("gene_spis"))
immune_og <- og[match(immune_dt[[1]], gene_spis)][!is.na(gene)]
immune_dt <- merge.data.table(immune_og, immune_dt, by = "gene_spis")
label_genes_immuno_2 <- structure(immune_dt[[3]], names = immune_dt[[2]])

# label other immune genes from stylophora
other_genes <- c(
  "Spis_XP_022802147_1" = "NLR",
  "Spis_XP_022803640_1" = "NLR",
  "Spis_XP_022789210_1" = "NFKB",
  "Spis_XP_022784447_1" = "RIG-I_C-RD/ResIII/Helicase_C/RdRP/",
  "Spis_XP_022784454_1" = "TBK1",
  "Spis_XP_022797663_1" = "TRAF6"
)
other_genes_dt <- og[match(names(other_genes), gene_spis)][!is.na(gene)]
other_genes_dt[, desc := other_genes[gene_spis]]
label_genes_immuno_3 <- c(
  structure(
    other_genes_dt[[3]], names = other_genes_dt[[2]]
  )
)

# label immune genes from crassostrea
genes_immuno <- c(
  "Cvig_LOC111107200" = "TBK1",
  "Cvig_LOC111107048" = "TBK1-like",
  "Cvig_LOC111107438" = "TBK1",
  "Cvig_LOC111109188" = "TBK1"
)
list_int_fixd <- dictionary_t2g(
  gtf_fn = file.path(
    "Crassostrea_virginica", "genome",
    "GCF_002022765.2_C_virginica-3.0_genomic.gtf"
  ),
  vector_to_fix = str_remove(names(genes_immuno), "Cvig_"),
  t2g = FALSE,
  transcript_field = "CDS",
  transcript_id = "protein_id",
  gene_id = "gene_id"
)
list_int_dt <- data.table(
  new = paste0("Cvig_", str_replace(list_int_fixd, "\\.", "_")),
  old = paste0("Cvig_", names(list_int_fixd))
)
list_int_dt[old == "Cvig_NA", old := new]
list_int_dt[, desc := genes_immuno[old]]
label_genes_immuno_4 <- structure(
  list_int_dt$desc, names = list_int_dt$new
)

# all genes to label
label_genes_immuno <- c(
  label_genes_immuno_4,
  label_genes_immuno_3,
  label_genes_immuno_2,
  label_genes_immuno_1
)
label_genes_v <- str_replace_all(label_genes_immuno, c(
  "\\." = "_",
  "IRF3/IRF4/IRF5/IRF6/IRF7/IRF8/IRF9" = "IRF3/4/5/6/7/8/9",
  "NFKB1/NFKB2/REL/RELA/RELB" = "NFKB1/2",
  "STAT5A/STAT5B/STAT6" = "STAT5/6",
  "Antiviral innate immune response receptor RIG-I" = "RIG-I",
  "DNA_damage_regulated_autophagy_modulator_1" = "DRAM1",
  "probable E3 ubiquitin-protein ligase RNF144A-B" = "RNF144A",
  "delta_like_canonical_Notch_ligand_1" = "DLL1",
  "calcium_sensing_receptor" = "CaSR",
  "guanylate_binding_protein_7" = "GBP7",
  "BAG_cochaperone_5" = "BAG5",
  "DnaJ_heat_shock_protein_family_Hsp40_member_B4" = "DNAJB4",
  "platelet glycoprotein V-like" = "GP5-like",
  "tetratricopeptide_repeat_domain_28" = "TTC28",
  "ubiquitin_conjugating_enzyme_E2_W" = "UBE2W",
  "ribosomal_protein_S6_kinase_A5" = "RPS6KA5",
  "heat_shock_protein_family_A_Hsp70_member_8" = "HSPA8",
  "caspase_recruitment_domain_family_member_11" = "CARD11",
  "BCL2_associated_X_apoptosis_regulator" = "BAX BCL2 associated X",
  "protein HEXIM-like isoform X2" = "HEXIM2",
  "lipopolysaccharide_binding_protein" = "LBP",
  "C_type_lectin_domain_family_4_member_C" = "CLEC4C"
))
names(label_genes_v) <- str_replace(names(label_genes_immuno), "\\.", "_")
label_genes_v <- label_genes_v[!grepl("uncharacterized", label_genes_v)]
```

cApUp

```{r}
# get results table - cApUp
contrast_capup <- c("treatment", "cApUp", "control")
res_capup <- results(dds, contrast = contrast_capup)
res_lfc_capup <- lfcShrink(dds, contrast = contrast_capup, type = "ashr")

# map cvig gene IDs to protein IDs
list_int_fixd <- dictionary_t2g(
  gtf_fn = file.path(
    "Crassostrea_virginica", "genome",
    "GCF_002022765.2_C_virginica-3.0_genomic.gtf"
  ),
  vector_to_fix = str_remove(rownames(res_lfc_capup), "Cvig_"),
  t2g = FALSE,
  transcript_field = "CDS",
  transcript_id = "protein_id",
  gene_id = "gene_id"
)
list_int_dt <- data.table(
  new = paste0("Cvig_", str_replace(list_int_fixd, "\\.", "_")),
  old = names(list_int_fixd)
)
list_int_dt[!is.na(old), old := paste0("Cvig_", old)]
list_int_dt[is.na(old), old := new]
res_lfc_capup$gene <- list_int_dt[match(rownames(res_lfc_capup), old), new]
names <- rownames(res_lfc_capup)
names[!is.na(res_lfc_capup$gene)] <- res_lfc_capup[
  !is.na(res_lfc_capup$gene),
]$gene
rownames(res_lfc_capup) <- names

# add gene lists annotations
res_lfc_capup$label <- "none"
res_lfc_capup[rownames(res_lfc_capup) %in% cglr_genes, ]$label <- "cGLR"
res_lfc_capup[rownames(res_lfc_capup) %in% sting_genes, ]$label <- "STING"
res_lfc_capup$label_gene <- label_genes_immuno_1[rownames(res_lfc_capup)]
res_lfc_capup[is.na(res_lfc_capup$label_gene), ]$label_gene <- ""

# plot cApUp
gp_capup_lab <- as.data.table(res_lfc_capup)[label != "none"][
  padj < padj_thr & abs(log2FoldChange) > lfc_thr
]
gp_capup_lab[, minuslog10padj := -1 * log10(padj)]
gp_capup_lab[, dir := ifelse(log2FoldChange > 0, "up", "down")]
gp_capup_vo <- ggplotVolcano(
  res_lfc_capup, padj_thr = padj_thr, lfc_thr = 1,
  lims_fc = c(-8, 8), lims_sig = c(NA, 80),
  sign_col = c("#F27333", "#725CA6"), size = 1.5, alpha = 0.8,
  title = sprintf("%s vs %s", contrast_capup[2], contrast_capup[3]),
  label_column = "label_gene", label_fc_thr = 0.5, label_size = 2.5,
  shape_column = "label", raster = TRUE
) +
  geom_point(
    data = gp_capup_lab, aes(shape = label, fill = dir),
    color = "black", size = 2
  ) +
  scale_shape_manual(values = c("STING" = 22, "cGLR" = 24, "none" = 20))
```

cGAMP

```{r}
# get results table - cGAMP
contrast_cgamp <- c("treatment", "cGAMP", "control")
res_cgamp <- results(dds, contrast = contrast_cgamp)
res_lfc_cgamp <- lfcShrink(dds, contrast = contrast_cgamp, type = "ashr")

# map cvig gene IDs to protein IDs
list_int_fixd <- dictionary_t2g(
  gtf_fn = file.path(
    "Crassostrea_virginica", "genome",
    "GCF_002022765.2_C_virginica-3.0_genomic.gtf"
  ),
  vector_to_fix = str_remove(rownames(res_lfc_cgamp), "Cvig_"),
  t2g = FALSE,
  transcript_field = "CDS",
  transcript_id = "protein_id",
  gene_id = "gene_id"
)
list_int_dt <- data.table(
  new = paste0("Cvig_", str_replace(list_int_fixd, "\\.", "_")),
  old = names(list_int_fixd)
)
list_int_dt[!is.na(old), old := paste0("Cvig_", old)]
list_int_dt[is.na(old), old := new]
res_lfc_cgamp$gene <- list_int_dt[match(rownames(res_lfc_cgamp), old), new]
names <- rownames(res_lfc_cgamp)
names[!is.na(res_lfc_cgamp$gene)] <- res_lfc_cgamp[
  !is.na(res_lfc_cgamp$gene),
]$gene
rownames(res_lfc_cgamp) <- names

# add gene lists annotations
res_lfc_cgamp$label <- "none"
res_lfc_cgamp[rownames(res_lfc_cgamp) %in% cglr_genes, ]$label <- "cGLR"
res_lfc_cgamp[rownames(res_lfc_cgamp) %in% sting_genes, ]$label <- "STING"
res_lfc_cgamp$label_gene <- label_genes_v[rownames(res_lfc_cgamp)]
res_lfc_cgamp[is.na(res_lfc_cgamp$label_gene), ]$label_gene <- ""

# plot cGAMP
gp_cgamp_lab <- as.data.table(res_lfc_cgamp)[label != "none"][
  padj < padj_thr & abs(log2FoldChange) > lfc_thr
]
gp_cgamp_lab[, minuslog10padj := -1 * log10(padj)]
gp_cgamp_lab[, dir := ifelse(log2FoldChange > 0, "up", "down")]
gp_cgamp_vo <- ggplotVolcano(
  res_lfc_cgamp, padj_thr = padj_thr, lfc_thr = 1,
  lims_fc = c(-5, 5), lims_sig = c(NA, 6),
  sign_col = c("#F27333", "#725CA6"), size = 1, alpha = 0.8,
  title = sprintf("%s vs %s", contrast_cgamp[2], contrast_cgamp[3]),
  raster = TRUE,
  label_column = "label_gene", label_fc_thr = 0.5, label_size = 2.5,
) +
  geom_point(
    data = gp_cgamp_lab, aes(shape = label, fill = dir),
    color = "black", size = 2
  ) +
  scale_shape_manual(values = c("STING" = 22, "cGLR" = 24, "none" = 20))

```

```{r}
gp_cgamp_vo + gp_capup_vo
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, sprintf("dds_res_ashr_cGAMP_padj%s.pdf", padj_thr)),
  width = 5, height = 6
)
gp_cgamp_vo
dev.off()

pdf(
  file.path(fig_dir, sprintf("dds_res_ashr_cApUp_padj%s.pdf", padj_thr)),
  width = 5, height = 6
)
gp_capup_vo
dev.off()
```

Save results

```{r}
# add gene annotations
res_list <- sapply(
  list(
    "cApUp" = res_lfc_capup,
    "cGAMP" = res_lfc_cgamp
  ),
  function(x) {
    dt <- as.data.table(x, keep.rownames = "gene")
    dt[is.na(pvalue), pvalue := 1]
    dt[is.na(padj), padj := 1]
    dt_ann <- merge.data.table(
      dt, gene_ann,
      by = "gene",
      all.x = TRUE
    )
    dt_ann[is.na(dt_ann)] <- ""
    dt_ann[, TF := tf_family != ""]
    dt_ann[, name := ""]
    setorder(dt_ann, padj, log2FoldChange)
    dt_ann
}, simplify = FALSE, USE.NAMES = FALSE)

# significant genes from gene lists of interest
dt_sign <- rbindlist(sapply(names(res_list), function(x) {
  res_list[[x]][
    padj < padj_thr & abs(log2FoldChange) > lfc_thr][
      order(-log2FoldChange)][, 1:8]
}, simplify = FALSE, USE.NAMES = TRUE), idcol = "treatment")
setorder(dt_sign, -log2FoldChange)

# significant TFs
dt_tfs <- rbindlist(sapply(names(res_list), function(x) {
  res_list[[x]][
    padj < padj_thr & abs(log2FoldChange) > lfc_thr & (TF == TRUE)][
      order(-log2FoldChange)][, 1:8]
}, simplify = FALSE, USE.NAMES = TRUE), idcol = "treatment")
setorder(dt_tfs, -log2FoldChange)

# save results to excel
res_list <- c(
  res_list,
  list(
    "significant_genes" = dt_sign,
    "TFs_significant genes" = dt_tfs
  )
)
wb <- createWorkbook()
for (i in names(res_list)) {
  message(i)
  addWorksheet(wb, sheetName = i)
  writeData(wb, sheet = i, res_list[[i]])
}
saveWorkbook(
  wb,
  file.path(res_dir, sprintf("dds_res_ashr.xlsx")),
  overwrite = TRUE
)
```

Differentially expressed genes in cApUp treatment

```{r}
res_lfc <- rbindlist(list(
  "cApUp" = as.data.table(res_lfc_capup, keep.rownames = "gene"),
  "cGAMP" = as.data.table(res_lfc_cgamp, keep.rownames = "gene")
), idcol = "treatment")
fwrite(res_lfc, file.path(res_dir, "dss_res.tsv"))

# up-regulated
res_up <- res_lfc[padj < padj_thr & log2FoldChange > 1]

# down-regulated
res_down <- res_lfc[padj < padj_thr & log2FoldChange < 1]
```

## GO

GO analysis on the upregulated genes.

```{r, eval=FALSE, include=FALSE}
# parse go annotations from eggNOG results
go_dat <- fread(
  file.path(ann_dir, "eggNOG", "MM_9o_a6n2t.emapper.annotations.tsv"),
  sep = "\t",
  fill = TRUE,
  skip = 4,
  header = TRUE
)
setnames(go_dat, "#query", "query")
go_dat[, .(query, GOs)]
go_map <- sapply(go_dat[, query], function(x) {
  gos <- go_dat[query == x, GOs]
  strsplit(gos, ",")[[1]]
}, simplify = FALSE, USE.NAMES = TRUE)
saveRDS(go_map, file.path(ann_dir, "gomap_cvir.RDS"))
```

```{r}
# differential analysis results
res_list <- list(
  up = res_up$gene,
  down = res_down$gene
)

# GO
go_map <- readRDS(file.path(ann_dir, "gomap_cvir.RDS"))
padj_thr <- 0.05
lfc_thr <- 1
tgalg <- "classic"
go_dir <- file.path(res_dir, sprintf("GO_%s_padj%s", tgalg, padj_thr))
dir.create(go_dir, recursive = TRUE, showWarnings = FALSE)

# calculate GO enrichment
go_dt <- rbindlist(sapply(names(res_list), function(i) {
    list_interest <- res_list[[i]]
    # map gene IDs to protein IDs
    list_int_fixd <- dictionary_t2g(
      gtf_fn = file.path(
        gen_dir, "GCF_002022765.2_C_virginica-3.0_genomic.gtf"
      ),
      vector_to_fix = str_remove(list_interest, "Cvig_"),
      t2g = FALSE,
      transcript_field = "CDS",
      transcript_id = "protein_id",
      gene_id = "gene_id"
    )
    list_int_prot <- list_int_fixd[!is.na(names(list_int_fixd))]
    list_int_prot <- paste0("Cvig_", list_int_prot)
    # enrichment
    setDT(topgofun(
      list_int_prot, go_map,
      tg_algorithm = tgalg,
      output_name = file.path(go_dir, "go"),
      name_geneset = i,
      ontology_set = c("MF", "CC", "BP"),
      firstSigNodes = 20
    ))
}, USE.NAMES = TRUE, simplify = FALSE), idcol = "group")
```

Save results

```{r}
# save
fwrite(
  go_dt,
  file.path(go_dir, sprintf("topGO.%s.padj%s.tsv", tgalg, padj_thr)),
  sep = "\t"
)
for (i in unique(go_dt$group)) {
  fwrite(
    go_dt[group == i & pval_adj < padj_thr],
    file.path(go_dir, sprintf("topGO.%s.padj%s.%s.tsv", tgalg, padj_thr, i)),
    sep = "\t"
  )
}
res_list <- sapply(unique(go_dt$group), function(i) {
  go_dt[group == i & pval_adj < padj_thr][, group := NULL]
}, simplify = FALSE, USE.NAMES = TRUE)
wb <- createWorkbook()
for (i in names(res_list)) {
  message(i)
  addWorksheet(wb, sheetName = i)
  writeData(wb, sheet = i, res_list[[i]])
}
saveWorkbook(
  wb,
  file.path(res_dir, sprintf("go_res_padj%s.xlsx", padj_thr)),
  overwrite = TRUE
)
```

Jaccard overlap between GO gene sets to select which GOs to show in the figure

```{r}
# select significant results
go_dt <- fread(
  file.path(go_dir, sprintf("topGO.%s.padj%s.tsv", tgalg, padj_thr))
)
go_dt[, label := sprintf("%s | %s", GO.ID, Term)]

# categories to consider
cats <- unique(go_dt[,
  .SD[order(pval_adj)][1:20][pval_adj < 0.01],
  .(group, ontology)
]$GO.ID)

# map genes to categories
go_map <- readRDS(file.path(ann_dir, "gomap_cvir.RDS"))
go_map_dt <- data.table(
  gene = rep(names(go_map), sapply(go_map, length)),
  category = unlist(go_map)
)

# calculate jaccard index for all categories pairs
jaccard <- function(x, y) {
  length(intersect(x, y)) / length(union(x, y))
}
jacc_mt <- matrix(nrow = length(cats), ncol = length(cats))
rownames(jacc_mt) <- colnames(jacc_mt) <- cats
for (x in cats) {
  for (y in cats) {
    jacc_mt[x, y] <- jaccard(
      go_map_dt[category == x]$gene,
      go_map_dt[category == y]$gene
    )
  }
}

# clusters
hc <- hclust(dist(jacc_mt, method = "maximum"), method = "ward.D2")
hm_mt <- jacc_mt[hc$order, hc$order]
rownames(hm_mt) <- go_dt[match(rownames(hm_mt), GO.ID)]$label

# heatmap
jacc_hm <- Heatmap(
  hm_mt, name = "jaccard",
  cluster_rows = FALSE, cluster_columns = FALSE,
  border = TRUE,
  rect_gp = gpar(col = "black", lwd = 1),
  heatmap_legend_param = list(
    direction = "horizontal", legend_width = unit(6, "cm")
  )
)

go_clusters <- cutree(hc, k = 45)
go_dt_clusters <- go_dt[group == "up"][
  match(names(go_clusters), GO.ID)][
    , cluster := go_clusters[GO.ID]]
cats_red <- go_dt_clusters[
  order(cluster, pval_adj)][
    , .SD[1], cluster]$GO.ID
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(
    fig_dir,
    sprintf("go%s_categories_jaccard_padj%s.pdf", tgalg, padj_thr)
  ),
  height = 16, width = 16,
  useDingbats = TRUE
)
draw(jacc_hm, heatmap_legend_side = "bottom")
dev.off()
```

Plot GO categories dotplot 

```{r}
plot_dt <- copy(go_dt)[GO.ID %in% cats_red]
plot_dt[, label := sprintf("%s | %s", GO.ID, Term)]
plot_dt[, go_id := factor(
  GO.ID,
  levels = unique(
    plot_dt[order(Significant)][["GO.ID"]]
  )
)]
plot_dt[, group := factor(group, levels = c("up", "down"))]
plot_dt[, gene_num := Significant][gene_num > 800, gene_num := 800]
go_vect <- structure(plot_dt$label, names = as.character(plot_dt$go_id))

gp <- ggplot(plot_dt, aes(group, go_id, size = gene_num, fill = pval_adj)) +
  geom_point(shape = 21) +
  labs(
    x = "", y = "GO term",
    size = "significant\ngenes",
    fill = "adjusted\np value"
  ) +
  scale_fill_distiller(
    breaks = c(0, 0.05), limits = c(0, 0.05), oob = scales::squish
  ) +
  scale_y_discrete(labels = go_vect, position = "right") +
  facet_grid(ontology ~ ., scales = "free", space = "free", switch = "y") +
  theme(
    strip.background = element_blank(),
    panel.grid.major = element_line(linewidth = 0.2),
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)
  )
gp
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, sprintf("go%s_plot_padj%s.pdf", tgalg, padj_thr)),
  height = 11,
  width = 8.6,
  useDingbats = TRUE
)
print(gp)
dev.off()
fwrite(
  plot_dt,
  file.path(res_dir, sprintf("go%s_plot_padj%s.tsv", tgalg, padj_thr)),
  sep = "\t"
)
```

```{r}
plot_cvir <- fread(file.path(
  res_dir, sprintf("go%s_plot_padj%s.tsv", tgalg, padj_thr)
))
plot_spis <- fread(file.path(
  "Stylophora_pistillata", "differential_expression", "results",
  sprintf("go%s_plot_padj%s.tsv", tgalg, padj_thr)
))
unique(plot_cvir[GO.ID %in% intersect(plot_spis$GO.ID, plot_cvir$GO.ID)][, .(ontology, label)])
```

## KEGG

Identify up or down regulated KEGG pathways

```{r, eval=FALSE, include=FALSE}
# parse KEGG annotations from eggNOG results
ko_dat <- fread(
  file.path(ann_dir, "eggNOG", "MM_9o_a6n2t.emapper.annotations.tsv"),
  sep = "\t",
  fill = TRUE,
  skip = 4,
  header = TRUE
)
setnames(ko_dat, "#query", "query")
ko_map <- ko_dat[!KEGG_Pathway %in% c("-", ""), .(query, KEGG_Pathway)]
ko_map <- tidyr::separate_rows(ko_map, KEGG_Pathway, sep = ",")
setDT(ko_map)
fwrite(
  ko_map,
  file.path(ann_dir, "Crassostrea_virginica_KEGGpathway_table.tsv"),
  sep = "\t"
)
```

```{r}
require(gage)

wdir <- getwd()
kegg_dir <- file.path(res_dir, sprintf("KEGG_padj%s", padj_thr))
dir.create(kegg_dir, showWarnings = FALSE)

# kegg pathways
ko <- fread(
  file.path(ann_dir, "Crassostrea_virginica_KEGGpathway_table.tsv"),
  header = TRUE
)
setnames(ko, c("gene", "ko"))
ko_ids <- unique(ko[[2]])
ko_list <- sapply(ko_ids, function(x) {
  ko[ko == x]$gene
}, USE.NAMES = TRUE, simplify = FALSE)


# map gene IDs to protein IDs
res_lfc_map <- dictionary_t2g(
  gtf_fn = file.path(
    gen_dir, "GCF_002022765.2_C_virginica-3.0_genomic.gtf"
  ),
  vector_to_fix = str_remove(res_lfc$gene, "Cvig_"),
  t2g = FALSE,
  transcript_field = "CDS",
  transcript_id = "protein_id",
  gene_id = "gene_id"
)
res_lfc_map <- res_lfc_map[!is.na(names(res_lfc_map))]
res_lfc_genes <- structure(
  paste0("Cvig_", res_lfc_map),
  names = paste0("Cvig_", names(res_lfc_map))
)

# gage
kegg_res <- gage(
  exprs = structure(
    res_lfc[match(names(res_lfc_genes), res_lfc$gene)]$log2FoldChange,
    names = unname(res_lfc_genes)
  ),
  gsets = ko_list,
  same.dir = FALSE,
  ref = NULL, samp = NULL
)

# parse gage results
kegg_dt <- as.data.table(kegg_res$greater, keep.rownames = "kegg.id")
kegg_dt <- kegg_dt[grepl("ko", kegg_dt$kegg.id)]

# mapping kegg ids to labels may take a while
kgs <- unique(kegg_dt$kegg.id)
kgl <- sapply(kgs, function(i) {
  tryCatch(
    keggGet(i)[[1]]$NAME,
    error = function(e) NA
  )
}, simplify = TRUE, USE.NAMES = TRUE)
kegg_dt[, kegg.lab := kgl[kegg.id]]
setcolorder(kegg_dt, c("kegg.id", "kegg.lab"))

# save all
fwrite(
  kegg_dt,
  file.path(kegg_dir, sprintf("kegg.results.padj%s.tsv", padj_thr)),
  sep = "\t"
)

kegg_dt[order(p.val)]
kegg_dt[kegg.id == "ko04024"]
```

Visualize KEGG pathways

```{r}
require(pathview)

kegg_dt <- fread(file.path(
  kegg_dir,
  sprintf("kegg.results.padj%s.tsv", padj_thr)
))

# top pathways
kegg_sig <- kegg_dt[q.val < padj_thr]
fwrite(
  kegg_sig,
  file.path(kegg_dir, sprintf("kegg.results.significant.padj%s.tsv", padj_thr)),
  sep = "\t"
)

res_sig <- kegg_sig[order(q.val)]
wb <- createWorkbook()
addWorksheet(wb, sheetName = "significant")
writeData(wb, sheet = "significant", res_sig)
saveWorkbook(
  wb,
  file.path(res_dir, sprintf("kegg_res_padj%s.xlsx", padj_thr)),
  overwrite = TRUE
)

# load kegg info
ko <- fread(file.path(
  ann_dir, "Crassostrea_virginica_KEGGpathway_table.tsv"
), header = FALSE)
setnames(ko, c("gene", "ko"))

# map gene IDs to protein IDs
ko_lfc <- res_lfc[, .(gene, log2FoldChange)]
res_lfc_map <- dictionary_t2g(
  gtf_fn = file.path(
    gen_dir, "GCF_002022765.2_C_virginica-3.0_genomic.gtf"
  ),
  vector_to_fix = str_remove(res_lfc$gene, "Cvig_"),
  t2g = FALSE,
  transcript_field = "CDS",
  transcript_id = "protein_id",
  gene_id = "gene_id"
)
res_lfc_map <- res_lfc_map[!is.na(names(res_lfc_map))]
res_lfc_genes <- structure(
  paste0("Cvig_", res_lfc_map),
  names = paste0("Cvig_", names(res_lfc_map))
)
ko_lfc[, gene_id := gene][, gene := res_lfc_genes[gene_id]]
ko_lfc <- ko_lfc[!is.na(gene)]

# add gene fold change to kegg info
ko_lfc <- merge.data.table(ko, ko_lfc, by = "gene", all.x = TRUE)
ko_lfc <- ko_lfc[!is.na(gene_id)]
ko_lfc[, gene_id := NULL]
ko_lfc <- unique(ko_lfc)
ko_lfc <- ko_lfc[, .(log2FoldChange = mean(log2FoldChange)), by = ko]

# plot top pathways
wdir <- getwd()
setwd(kegg_dir)
for (k in unique(kegg_sig$kegg.id)) {
  lab <- kegg_sig[match(k, kegg.id)]$kegg.lab
  message("")
  message(k, " | ", lab)
  message("")
  tryCatch({
    pathview(
      gene.data = structure(ko_lfc$log2FoldChange, names = ko_lfc$ko),
      pathway.id = k,
      gene.idtype = "KEGG",
      species = "ko",
      limit = list(gene = 1, cpd = 1),
      both.dirs = TRUE,
      high = list(gene = "chartreuse"),
      out.suffix = x
    )
  }, error = function(e) warning("Failed plotting pathway for ", k))
}
setwd(wdir)
```

Dotplot

```{r}
kegg_dt <- fread(file.path(kegg_dir, "kegg.results.tsv"))
kegg_sig <- fread(
  file.path(kegg_dir, sprintf("kegg.results.significant.padj%s.tsv", padj_thr)),
  sep = "\t"
)

# plot
plot_dt <- kegg_dt[order(q.val)][kegg.id %in% kegg_sig[q.val < 0.01]$kegg.id][1:20]
plot_dt[, label := sprintf("%s | %s", kegg.id, kegg.lab)]
plot_dt[, kegg_id := factor(
  kegg.id,
  levels = unique(
    plot_dt[order(stat.mean)][["kegg.id"]]
  )
)]
#plot_dt[, change := stat.mean][change > 800, change := 800]
plot_dt[, group := "cApUp"]
go_vect <- structure(plot_dt$label, names = as.character(plot_dt$kegg_id))

gp <- ggplot(plot_dt, aes(group, kegg_id, size = stat.mean, fill = q.val)) +
  geom_point(shape = 21) +
  labs(
    x = "", y = "KEGG pathway",
    size = "expression\nchange",
    fill = "adjusted\np value"
  ) +
  scale_fill_gradient(
    low = "#725CA6", high = "#ffffff",
    breaks = c(0, 0.05), limits = c(0, 0.05), oob = scales::squish
  ) +
  scale_y_discrete(labels = go_vect, position = "right") +
  theme(
    strip.background = element_blank(),
    panel.grid.major = element_line(linewidth = 0.2),
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)
  )
gp
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, sprintf("kegg_plot_padj%s.pdf", 0.01)),
  height = 6,
  width = 8.5,
  useDingbats = TRUE
)
print(gp)
dev.off()
fwrite(
  plot_dt,
  file.path(res_dir, sprintf("kegg_plot_padj%s.tsv", 0.01)),
  sep = "\t"
)
```

```{r}
plot_cvir <- fread(file.path(
  res_dir, sprintf("kegg_plot_padj%s.tsv", padj_thr)
))
plot_spis <- fread(file.path(
  "Stylophora_pistillata", "differential_expression", "results",
  sprintf("kegg_plot_padj%s.tsv", padj_thr)
))
unique(plot_cvir[
  kegg.id %in% intersect(plot_spis$kegg.id, plot_cvir$kegg.id)
][, .(label)])
```

## Session info

```{r}
sessionInfo()
```